### PR TITLE
[feature] Add persistent and scheduled mass upgrades

### LINKED ISSUE #379: [feature] Persistent Mass Upgrades
Currently, firmware upgrades in OpenWISP happen immediately via Celery tasks.
If a device is offline at the moment of glory, the task fails, and someone has to manually retry it.

This becomes a nightmare in large deployments: admins can't just schedule upgrades and walk away.

What our users want: **mass upgrades that patiently wait for devices to come back online and then do their job automatically**.

## Describe the solution you'd like

Image

Introduce support for **persistent upgrade tasks** that stick around until offline devices finally wake up and get upgraded.

This is only for **mass upgrade operations** for now; one-device-at-a-time upgrades keep their current behavior.

The system should ask users whether they want the mass upgrade to retry indefinitely, with the default option suggested as enabled (checked by default in both the admin interface and REST API).
This will likely require a new boolean database column, for example `persistent` (yes/no), which should be visible in the admin and REST API lists for mass upgrade operations.
Once a mass upgrade operation has been launched, this field **must not** be changeable.

Some ideas and guidelines for contributors (but feel free to do your own research and come up with better ideas if deemed proper):

* **Persistent task records**

  * Use the existing `UpgradeOperation` model for most of the data.
  * Add `retry_count` and optionally `scheduled_time` to track when the next retry is due.
  * These tasks must survive system restarts and database migrations.

* **Device online detection**

  * Prefer using the `health_status_changed` signal from OpenWISP Monitoring (mock it in CI/testing).
  * Fallback: periodic retries with **randomized exponential backoff** (configurable, max once every 12 hours).

* **Retry strategy**

  * Randomized exponential backoff with indefinite retries (up to a practical high limit).
  * For long-running mass upgrades (default 2 months), notify the admin via `generic_notification` about devices still waiting for upgrade.
  * Reminders repeat periodically until the admin cancels the operation or all devices are upgraded.

* **Integration with Celery**

  * Existing upgrade operations already run as Celery tasks.
  * Introduce a new task (or tasks) to “wake up” pending upgrades for mass upgrades.
  * Randomized retry delays prevent all upgrades from running simultaneously and overloading the system.

* **Failure handling & notifications**

  * Failures that need human attention: devices offline too long, upgrade errors, checksum issues.
  * Use `generic_notification` to inform the admins.
  * Failed upgrades require manual review; automatic retries after fixes are optional.

* **Edge cases**

  * If `health_status_changed` signal trigges the wake up of an upgrade opration while the upgrade operation is seemingly already woken up → ignore and do nothing.
  * Only one upgrade per device is allowed; queue order and conflicts area already handled and no changes should be needed.
  * No rollback support post-flashing as it's not technically possible and firmware conflicts already managed by existing logic.

* **Scalability**

  * No hard-coded limit on pending tasks; system scales with available workers.
  * Randomized backoff prevents broker/database overload.
  * No batching required; retries remain random for now.

* **Metrics & observability**

  * Track retry counts for upgrade oprations (visible in admin/REST API).
  * Failures already handled by existing logic.
  * Minimal admin UI and REST API exposure for filtering pending upgrade operations that are going to be retried.

The nitty-gritty of how to schedule retries, detect online devices, and handle notifications is left for contributors to explore.
**Safety and reliability is key**.

## Describe alternatives you've considered

* Retry failed tasks immediately → nope, offline devices would never get upgraded.
* Manual re-triggering → scales about as well as a greased pig on ice.

## Additional context

* This feature is all about making mass upgrades less of a headache and more like magic.
* Contributors are encouraged to explore different strategies for persistent queuing, backoff logic, and Celery integration.
* Consider edge cases like very long offline periods, multiple upgrades queued for the same devices, and safe notification handling.

# Constraints

* Test coverage **must not decrease**.
* Documentation needs to be updated to include this new feature, including updating any existing screenshots that may change after implementation.

### LINKED ISSUE #380: [feature] Scheduled Mass Upgrades
## Problem statement

Currently, mass firmware upgrades start immediately after the upgrade operation is created. This is problematic in production environments where devices must remain available during business hours. Network operators need a way to plan upgrades during low usage windows, for example overnight or during scheduled maintenance periods, without manual intervention at execution time.

## Describe the solution you'd like

```mermaid
flowchart TD
    A[User creates mass upgrade] --> B{Schedule datetime provided?}

    B -- No --> C[Execute immediately, existing behavior]
    C --> D[Mass upgrade runs via Celery tasks]
    D --> E[Upgrade completes]
    E --> F[Send generic notification]

    B -- Yes --> G[Validate scheduled datetime]
    G -- Invalid --> H[Reject with validation error]
    G -- Valid --> I[Create operation with status SCHEDULED and store UTC datetime]

    I --> J[Periodic scheduler via Celery Beat every minute]
    J --> K{Upgrade due?}

    K -- No --> J
    K -- Yes --> L[Transition status from SCHEDULED to RUNNING]

    L --> M[Runtime validation of permissions firmware and devices]

    M -- All invalid --> N[Cancel operation with FAILED status and error]
    N --> O[Send generic notification for failure]

    M -- Some or all valid --> P[Execute mass upgrade using existing logic]
    P --> Q[Upgrade completes]
    Q --> R[Send generic notification for completion]
```

Introduce support for **scheduled mass upgrades**, allowing a mass upgrade operation to start at a user defined datetime in the future.

Scheduled upgrades must be reliable, persistent across restarts, and clearly communicate timing semantics and execution state to users.

### UI

* On the existing mass upgrade confirmation page, allow users to optionally schedule the upgrade for a future datetime.
* By default, the upgrade executes immediately unless a future datetime is explicitly set.
* The scheduled datetime must:
  * be in the future
  * respect a configurable minimum delay (eg 10 minutes, to avoid accidental scheduling)
  * not exceed a configurable maximum horizon (eg default 6 months)
* Invalid scheduling values must be rejected with a validation error.
* Only upgrades in `scheduled` status are editable. Once execution starts, the operation becomes read only.
* Timezone handling must be explicit:
  * User input should be interpreted in the user browser timezone (or another clearly justified UX choice).
  * Datetimes must be stored in UTC.
  * The server preferred timezone should be clearly indicated in the UI.

### Backend / REST API

* Extend the REST API to allow creating, editing, and cancelling scheduled mass upgrade operations.
* Full feature parity between Django admin and REST API is required.
* The mass upgrade status model must be extended to include a `scheduled` state.
* Store the scheduled execution datetime in a nullable `scheduled_at` field.
* Persist any batch execution options that are currently only passed to the immediate Celery task, especially `firmwareless`, so scheduled upgrades can execute later with the same user choices.
* Status transitions are expected to include at least:
  * scheduled → running
  * scheduled → cancelled
  * scheduled → failed
* An operation is considered editable **only while its status is `scheduled`**.
* The exact editable fields while scheduled should be defined explicitly. A reasonable default is:
  * editable: `scheduled_at`, group/location targeting fields, `persistent`, and `firmwareless`
  * not editable: build/firmware target and `upgrade_options`
* The admin list, detail view, and REST API responses must clearly expose:
  * whether an operation is scheduled or immediate
  * the scheduled execution datetime (UTC)

### Execution model

* Firmware upgrades already run asynchronously using Celery tasks.
* Scheduled execution must:
  * survive server restarts and worker crashes
  * avoid using [Celery eta / countdown](https://docs.celeryq.dev/en/latest/userguide/calling.html#eta-and-countdown) due to well-known limitations for executing tasks in the far future
  * schedule **one task per mass upgrade operation**, not per device
* Using Celery Beat with a periodic task that scans and executes due scheduled upgrades is a valid and recommended approach.
* Contributors are expected to propose and justify the final scheduling mechanism.

### Runtime validation and failure handling

* Devices, permissions, firmware availability, and other preconditions must be re-evaluated **at execution time**
* If all target devices become invalid at execution time, the upgrade must be cancelled with a visible error state.
* Partial changes (eg some devices removed) are acceptable and should be handled by existing upgrader logic.
* Conflicting mass upgrade operations must not be allowed:
  * If a scheduled mass upgrade exists, users must not be able to create another conflicting mass upgrade (scheduled or immediate).
  * Contributors should verify this using tests and adjust the implementation only if needed (TDD approach).
  * Conflict checks should also consider persistent `pending` per-device operations, because they are still active upgrade attempts waiting for retry.
  * We should explicitly define what constitutes a conflict. One proposed starting point is an active batch operation (`idle`, `scheduled`, or `in-progress`) targeting the same category/device population, plus any active per-device operation in `in-progress` or `pending`.

### Notifications

* A `generic_notification` must be sent when a mass upgrade starts.
* A `generic_notification` must be sent when any mass upgrade completes (if not already implemented).
* A `generic_notification` must be sent if a scheduled upgrade fails runtime validation before execution.
* Notifications are delivered to all organization administrators and superusers via the existing notifications module, this is already implemented and no further action is required, just fire the `generic_notification` when due.

### Non functional requirements

* Test coverage **must not decrease**.
* Time dependent logic may be tested using mocking; tests currently run with `CELERY_EAGER=True`.
* Tests should cover scheduled + persistent composition: a scheduled mass upgrade starts later, upgrades online devices, moves offline devices to pending, and retries them later.
* Documentation must be updated to describe the new feature, including updating any affected screenshots.
* Provide a short example usage video suitable for YouTube, to be showcased on the website and in the documentation.

## Describe alternatives you've considered

* Using external cron jobs or manual scripts to delay upgrade creation. This is error prone, not user friendly, and bypasses OpenWISP’s auditability and permissions model.
* Relying solely on server side timezone assumptions without exposing them in the UI, which risks misconfiguration.

## Additional context

Scheduled execution is a common requirement in network management systems and aligns with real world operational practices.

## Constraints

* Test coverage **must not decrease**.
* Basic browser tests for UI related features are **required**.
* Documentation needs to be updated to include this new feature, including updating any existing screenshots that may change after implementation.
* We also need a short example usage video for YouTube that we can showcase on the website/documentation.

### PR DESCRIPTION
**Pull request by Eeshu-Yadav**
## Checklist

- [x] I have read the OpenWISP Contributing Guidelines and Anti AI Spam Policy.
- [x] I have manually tested the changes proposed in this pull request.
- [x] I have written new test cases for new code and/or updated existing tests for changes to existing code.
- [x] I have updated the documentation.

## Reference to Existing Issue

Closes #379.
Closes #380.

## Description of Changes

These are the two mass-upgrade features I worked on for GSoC 2026.

**Persistent mass upgrades (#379).** Until now, a device that was offline when a mass upgrade ran would just fail. Now the upgrade keeps it in a `pending` state and retries it in the background, with backoff, until it comes back online or the operation is cancelled. It is on by default, and each device operation inherits the setting from its batch. The batch detail page lists the pending devices and their retry state, and organization admins get notified once a persistent operation finally succeeds or is given up on.

**Scheduled mass upgrades (#380).** The confirmation page has a new "Scheduled at" field. Leave it empty and the upgrade runs immediately, like before; pick a future date and time and it launches itself then, which is handy for maintenance windows. You enter the time in your own timezone and it is stored in UTC, so there is no offset maths to do by hand, and times that don't exist or are ambiguous because of a DST change are rejected instead of being saved wrong. A Celery Beat task launches batches once they are due. While a batch is still scheduled you can reschedule or cancel it from its detail page, and if you try to create one that overlaps an already scheduled or running batch for the same firmware category, it is blocked and the clashing operation is named in the error. The REST API also gets a `scheduled_at` field and dedicated reschedule and cancel endpoints.

Both are covered by unit, API and Selenium tests, and the documentation is updated.

## Screenshot

Scheduled upgrade — confirmation page with the "Scheduled at" field:

Confirm

Scheduled batch detail with reschedule and cancel:

Detail

Conflict prevention when an overlapping batch already exists:

Conflict

Walkthrough:

Walkthrough

Persistent upgrade — pending operation retried in the background:

Pending operation

Notification when a persistent operation finishes:

Notification

### CODE REVIEW

File: openwisp_firmware_upgrader/base/models.py [hunk 1]
```diff
@@ -611,13 +713,302 @@ def clean(self):
...
+                .filter(pk=self.pk, status__in=("idle", "in-progress"))
+                .first()
+            )
+            if locked is None or locked.upgradeoperation_set.exists():
+                return
+            self.status = "in-progress"
+            self.save(update_fields=["status"])
+            self.upgrade_related_devices()
+            if firmwareless:
+                self.upgrade_firmwareless_devices()
```
[the remark this answers is not in the document]
Comment (Eeshu-Yadav) [reply openwisp_firmware_upgrader/base/models.py:966]: Not stuck: `_recover_orphaned_launches` plus the next eligibility scan converge a zero-child batch to `failed` within ~SCHEDULE_LAUNCH_TIMEOUT + one beat cycle, so no change needed.

File: openwisp_firmware_upgrader/base/models.py [hunk 2]
```diff
@@ -794,37 +1242,228 @@ def calculate_and_update_status(self):
...
+                except ObjectDoesNotExist:
+                    continue
+                if batch.status != "scheduled" or batch.scheduled_at > now:
+                    continue
+                lock_category(batch.build.category_id)
+                result = cls.dry_run(
+                    build=batch.build, group=batch.group, location=batch.location
+                )
+                eligible = result["device_firmwares"].exists() or (
+                    batch.firmwareless and result["devices"].exists()
```
[the remark this answers is not in the document]
Comment (Eeshu-Yadav) [reply openwisp_firmware_upgrader/base/models.py:1382]: Good catch, this is a real safety concern. The proper fix is a deletion-time guard that fails and notifies pending batches referencing the group/location; I'd rather land that as a focused follow-up than fold a behavioural change into this PR.

File: openwisp_firmware_upgrader/base/models.py [hunk 3]
```diff
@@ -794,37 +1242,228 @@ def calculate_and_update_status(self):
...
+        Reverts scheduled batches stuck ``in-progress`` with no children past
+        the launch timeout back to ``scheduled`` for a later scan to redispatch.
+        """
+        stale = now - timedelta(seconds=app_settings.SCHEDULE_LAUNCH_TIMEOUT)
+        cls.objects.filter(
+            status="in-progress",
+            scheduled_at__isnull=False,
+            modified__lte=stale,
+            upgradeoperation__isnull=True,
+        ).update(status="scheduled")
```
[the remark this answers is not in the document]
Comment (Eeshu-Yadav) [reply openwisp_firmware_upgrader/base/models.py:1427]: Double-dispatch is already prevented: `select_for_update` serialises the recovery UPDATE with `upgrade()`, and the `upgradeoperation_set.exists()` idempotency check makes a redispatched task a no-op, so it self-heals. Raising SCHEDULE_LAUNCH_TIMEOUT is the knob if queue latency stays high.

[the remark this answers is not in the document]
Comment (Eeshu-Yadav) [reply openwisp_firmware_upgrader/base/models.py:1427]: Double-dispatch is already prevented: `select_for_update` serialises the recovery UPDATE with `upgrade()` and the `upgradeoperation_set.exists()` check makes a redispatch a no-op, so it self-heals; leaving as-is.

File: openwisp_firmware_upgrader/base/models.py [hunk 4]
```diff
@@ -794,37 +1242,228 @@ def calculate_and_update_status(self):
...
+        Claims every scheduled mass upgrade whose ``scheduled_at`` has elapsed
+        and dispatches it, failing those no longer eligible or in conflict.
+        """
+        now = timezone.now()
+        cls._recover_orphaned_launches(now)
+        due_ids = list(
+            cls.objects.filter(status="scheduled", scheduled_at__lte=now).values_list(
+                "pk", flat=True
+            )
+        )
```
[the remark this answers is not in the document]
Comment (Eeshu-Yadav) [reply openwisp_firmware_upgrader/base/models.py:1364]: Intentional asymmetry: the due-scan processes a handful of admin-created batches per tick (a UUID list is negligible), while the pending-retry scanner streams because it is per-device and reaches thousands.

File: openwisp_firmware_upgrader/templates/admin/firmware_upgrader/batch_upgrade_operation_change_form.html [hunk 5]
```diff
@@ -12,6 +12,92 @@
...
+      <button type="button" id="batch-reschedule-cancel" class="button">{% trans "Discard" %}</button>
+    </div>
+  </div>
+  {% endif %}
+</section>
+{% endif %}
+
+{% if batch.status != "scheduled" %}
 <!-- Upgrade Operations Section -->
 <h1 class="upgrade-operations-title">{% trans "Upgrade Operations" %}</h1>
```
[the remark this answers is not in the document]
Comment (Eeshu-Yadav) [reply openwisp_firmware_upgrader/templates/admin/firmware_upgrader/batch_upgrade_operation_change_form.html]: By design: there are no operations to show while the batch is scheduled and the reschedule form uses that space. The only gap is keeping this exact form open across the beat trigger, which a reload covers.

File: openwisp_firmware_upgrader/api/views.py [hunk 6]
```diff
@@ -45,6 +48,38 @@
...
+                "error": openapi.Schema(
+                    type=openapi.TYPE_STRING, description=_("Error message")
+                )
+            },
+        ),
+    )
+
+
+def _error_response(message, status_code):
+    return Response({"error": message}, status=status_code)
```
[the remark this answers is not in the document]
Comment (Eeshu-Yadav) [reply openwisp_firmware_upgrader/api/views.py:80]: False positive: the returned messages are curated translated ValidationError strings, and the catch-all returns a generic message with the detail logged server-side. No raw exception or stack text is exposed.

**Comment by nemesifier** [comment]
@coderabbitai revise adherence to the rules and best practices described in AGENTS.md, can you find any violation?

File: openwisp_firmware_upgrader/settings.py [hunk 7]
```diff
@@ -24,6 +24,71 @@
...
+)
+PERSISTENT_RETRY_OPTIONS.update(
+    getattr(settings, "OPENWISP_FIRMWARE_UPGRADER_PERSISTENT_RETRY_OPTIONS", {})
+)
+if (
+    PERSISTENT_RETRY_OPTIONS["base_delay"] <= 0
+    or PERSISTENT_RETRY_OPTIONS["max_delay"] <= 0
+    or PERSISTENT_RETRY_OPTIONS["multiplier"] < 1
+    or not 0 <= PERSISTENT_RETRY_OPTIONS["jitter"] < 1
+):
```
[the remark this answers is not in the document]
Comment (Eeshu-Yadav) [reply openwisp_firmware_upgrader/settings.py:46]: Fixed. Added `dispatch_jitter > 0` and `signal_jitter >= 0` to the same validation block, with reload tests for both. `dispatch_jitter` has to be strictly positive because it also doubles as the re-claim marker (`next_retry_at = now + dispatch_jitter`); at zero the marker lands at the current time and the next beat re-dispatches, so rejecting only negatives isn't enough there.

File: openwisp_firmware_upgrader/base/models.py [hunk 8]
```diff
@@ -611,13 +713,307 @@ def clean(self):
...
+                    "pk": existing.pk,
+                    "status": existing.get_status_display(),
+                    "when": when,
+                }
+        UpgradeOperation = load_model("UpgradeOperation")
+        device_filter = models.Q(
+            device_id__in=self.build._find_related_device_firmwares(
+                group=self.group, location=self.location
+            ).values_list("device_id", flat=True)
+        )
```
[the remark this answers is not in the document]
Comment (Eeshu-Yadav) [reply openwisp_firmware_upgrader/base/models.py:931]: You're right that the category lock only guards the batch side and the standalone `create_upgrade_operation()` path doesn't take it. The window is real, but the outcome isn't a double flash: the existing per-device guard in `upgrade()` (it aborts when the device already has an in-progress/pending operation) stops one of the two before anything reaches the device, so at worst the batch records one aborted child. Fully serializing the two paths means adding category locking / conflict detection to the single-device flow, which sits outside #379/#380 — I'd rather do that as a focused follow-up than widen this PR.

File: openwisp_firmware_upgrader/base/models.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Eeshu-Yadav) [reply openwisp_firmware_upgrader/base/models.py:1427]: Agreed the children stay idempotent, and you're right that a too-low `SCHEDULE_LAUNCH_TIMEOUT` would keep resetting a batch before its task starts. I've documented the setting in `docs/user/settings.rst`, including the requirement to keep it above the worst-case dispatch-to-worker latency. The default of 300s sits comfortably above normal launch latency.
